src/fact_emitter.rs

```diff
@@ -0,0 +1,575 @@
+
+        for (idx, s) in bb.statements.iter().enumerate() {
+            // TODO: emit `node_text` per statement, but that string could be
+            // collected at the parser level
+
+            match s {
+                Statement::Assign(place, expr) => {
+                    // Emit facts about the assignment LHS
+                    let (lhs_ty, lhs_origins) = self.ty_and_origins_of_place(place);
+
+                    // Assignments clear all origins in the type
+                    for origin in &lhs_origins {
+                        facts
+                            .clear_origin
+                            .push((origin.clone(), node_at(&bb.name, idx)));
+                    }
+
+                    if !lhs_ty.is_ref() {
+                        // Assignments to non-references invalidate loans borrowing from them.
+                        //
+                        // TODO: handle assignments to fields and loans taken on subsets of
+                        // their paths. Until then: only support invalidations on assignments
+                        // to complete places.
+                        //
+                        if let Some(loans) = self.loans.get(place) {
+                            for (origin, _location) in loans {
+                                // TODO: if the `location` where the loan was issued can't
+                                // reach the current location, there is no need to emit
+                                // the invalidation
+                                facts
+                                    .invalidate_origin
+                                    .push((origin.clone(), node_at(&bb.name, idx)));
+                            }
+                        }
+                    }
+
+                    // Emit facts about the assignment RHS: evaluate the `expr`
+                    self.emit_expr_facts(bb, idx, expr, facts);
+
+                    // Relate the LHS and RHS tys
+                    self.emit_subset_facts(bb, idx, &lhs_ty, expr, facts);
+                }
+
+                Statement::Drop(expr) => {
+                    // Evaluate the `expr`
+                    self.emit_expr_facts(bb, idx, expr, facts);
+
+                    // TODO: is there something more to do because we're in a "drop" ?
+                }
+            }
+        }
+    }
+
+    fn emit_expr_facts(&self, bb: &BasicBlock, idx: usize, expr: &Expr, facts: &mut Facts) {
+        match expr {
+            Expr::Access { kind, place } => {
+                match kind {
+                    // Borrowing clears its origin: it's issuing a fresh origin of the same name
+                    AccessKind::Borrow(origin) | AccessKind::BorrowMut(origin) => {
+                        facts
+                            .clear_origin
+                            .push((origin.into(), node_at(&bb.name, idx)));
+
+                        if matches!(kind, AccessKind::BorrowMut(_)) {
```

Comment:
Shared borrows should also invalidate mutable ones of the same place. This requires tracking the kind of a borrow in `loans`. You allude to this in one of your comments in `FactEmitter::new`. Not sure how this extends to types with nested origins though.